File: schema-api-gwt/src/main/java/dev/sayaya/handbook/usecase/TypeRepository.java
package dev.sayaya.handbook.usecase;

import dev.sayaya.handbook.domain.LayoutPeriod;
import dev.sayaya.handbook.domain.TypeValue;
import dev.sayaya.rx.Observable;
import jsinterop.base.JsPropertyMap;

import java.util.List;
import java.util.Set;

/**
 * 타입 정보를 관리하는 저장소 인터페이스.
 */
public interface TypeRepository {
    Observable<Set<TypeValue>> list(LayoutPeriod period);
    Observable<Set<TypeValue>> save(Set<TypeValue> types);
    Observable<Set<TypeValue>> patch(List<JsPropertyMap<?>> patches);
    Observable<Void> delete(Set<TypeValue> types);
    Observable<Set<TypeValue>> versions(String typeId);
}

File: persist-type/src/main/kotlin/dev/sayaya/handbook/usecase/TypeRepository.kt
package dev.sayaya.handbook.usecase

import dev.sayaya.handbook.domain.Type
import dev.sayaya.handbook.domain.TypePatch
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Instant
import java.util.*

/** 타입 영속화 포트. */
interface TypeRepository {
    fun findByWorkspaceAndPeriod(workspace: UUID, effectDateTime: Instant, expireDateTime: Instant): Flux<Type>
    fun save(workspace: UUID, types: List<Type>): Flux<Type>
    fun patch(workspace: UUID, patches: List<TypePatch>): Flux<Type>
    fun delete(workspace: UUID, types: List<Type>): Mono<Void>
}
